### PR TITLE
Feature/add bsm helper

### PR DESCRIPTION
**Pull request by lewisid**
# PR Details
## Description

Since the ids are meant to be read as hexadecimal values instead of decimal values, the MobilityPath and MobilityOperation nodes were being sent the 10-digit decimal value as opposed to the appropriate maximum 8-digit hexadecimal values. To rectify this, I created a new node in carma-utils that would convert the ids broadcasted from the bsm_outbound topic to hexadecimal strings and added the method to the mobilitypath_publisher and sci_strategic_plugin nodes in carma-platform

## Related Issue

https://github.com/usdot-fhwa-stol/carma-platform/issues/1519
## Motivation and Context

It was discovered that both the MobilityPath and MobilityOperation header hostBSMId were incorrect. The BSM id maximum allowed length is 8, however both MobilityPath and MobilityOperation are sending 10 digits BSM id to the cpp_message node in carma-messenger. This means that the cpp_message node will ignore the 10 digits bsm id and replace it with the 8 digit default id "00000000".

## How Has This Been Tested?

This has been both unit tested using gtest/gmock and integration tested in the Ford Fusion passenger vehicle.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: bsm_helper/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Copyright (C) 2020-2021 LEIDOS.
```
Comment (msmcconnell) [inline bsm_helper/CMakeLists.txt:1]: Should just be 2021 since this file didn't exist in 2020

File: bsm_helper/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+## Compile as C++14, supported in ROS Noetic and newer
+add_compile_options(-std=c++14)
+
+## Find catkin macros and libraries
+## if COMPONENTS list like find_package(catkin REQUIRED COMPONENTS xyz)
+## is used, also find other catkin packages
+find_package(catkin REQUIRED COMPONENTS
+  cav_msgs
+  roscpp
```
Comment (msmcconnell) [inline bsm_helper/CMakeLists.txt:26]: I don't see usage of roscpp or cav_msgs anywhere in this package. Please remove all references to unused dependencies.

File: bsm_helper/include/bsm_helper/bsm_helper.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace BSMHelper {
+
+    class BSMHelper
+    {
+        public:
+
+            /**
+             * @brief Converts bsm id from uint8_t vector to hex string
+             * @param id The incoming bsm id
```
Comment (msmcconnell) [inline bsm_helper/include/bsm_helper/bsm_helper.h:18]: can you add a bit of description about the encoding (how the input array needs to be structured)

File: bsm_helper/src/bsm_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#include "bsm_helper/bsm_helper.h"
+
+namespace BSMHelper{
+
+
+    std::string BSMHelper::bsmIDtoString(std::vector<uint8_t, std::allocator<uint8_t>> id)
```
Comment (msmcconnell) [inline bsm_helper/src/bsm_helper.cpp:6]: Please add a size check to this method which verifies the input array has a length that fits within the requirements of a BSM. Throw an exception in this case.

File: bsm_helper/src/bsm_helper.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        for(size_t i =0; i < id.size();i++)
+        {
+            id_test.append(std::to_string(id.at(i)));
+        }
+
+        std::stringstream int_id(id_test);
+
+        unsigned long id_num;
+
+        int_id >> id_num;
```
Comment (msmcconnell) [inline bsm_helper/src/bsm_helper.cpp:20]: Converting to a string and then pushing into a long just to convert back to a string seems pretty round about. Couldn't you just push each byte directly into the "unsigned long" using bit wise operators?

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: bsm_helper/CMakeLists.txt (hunk 1 above)
Comment (msmcconnell) [reply bsm_helper/CMakeLists.txt:1]: check all files

File: bsm_helper/include/bsm_helper/bsm_helper.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#include <string>
```
Comment (msmcconnell) [inline bsm_helper/include/bsm_helper/bsm_helper.h:1]: Almost missed this,

You need to add a 
#pragma once 
to the top of this file as an include guard. 
Also all source code files though have the license comment

File: bsm_helper/src/bsm_helper.cpp [hunk 7]
```diff
@@ -1,23 +1,30 @@
...
 namespace BSMHelper{
 
 
     std::string BSMHelper::bsmIDtoString(std::vector<uint8_t, std::allocator<uint8_t>> id)
     {
-
-        std::string id_test;
-
-        for(size_t i =0; i < id.size();i++)
+        if(id.size() > 4)
```
Comment (msmcconnell) [inline bsm_helper/src/bsm_helper.cpp:18]: Should this be !=
Is a 3 element input valid?

File: bsm_helper/src/bsm_helper.cpp [hunk 8]
```diff
@@ -1,23 +1,30 @@
...
 
-        std::stringstream int_id(id_test);
-
-        unsigned long id_num;
-
-        int_id >> id_num;
+        unsigned long id_num = 0;
+        for(size_t i = 0; i < id.size(); i++)
+        {
+          id_num = (id_num * pow(10, 2)) + id.at(i) ;
```
Comment (msmcconnell) [inline bsm_helper/src/bsm_helper.cpp:26]: I think this method is prone to unit conversion errors. The pow function returns a double which means id_num might be upcast first which could result in a bit flip. At least that's what my intuition tells me. I thick bitwise operators are the better choice here

```
for(...)
{
   id_num = (id_num << 8) | id.at(i);
}
```

See
https://www.geeksforgeeks.org/bitwise-operators-in-c-cpp/

**Review by msmcconnell — CHANGES_REQUESTED** [review]

**Review by dan-du-car — APPROVED** [review]

File: bsm_helper/src/bsm_helper.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            throw std::invalid_argument("Invalid id value");
+        }
+
+        unsigned long id_num = 0;
+        for(size_t i = 0; i < id.size(); i++)
+        {     
+
+            id_num = (id_num << 8) | id.at(i);
+            std::cout << id_num << std::endl;
+            std::cout<< std::hex << static_cast<int>(id.at(i)) << std::endl;
```
Comment (msmcconnell) [inline bsm_helper/src/bsm_helper.cpp:29]: These prints should either be removed or made into ros debug prints

**Review by msmcconnell — CHANGES_REQUESTED** [review]

**Review by msmcconnell — APPROVED** [review]

**Pull request merged**